DQMServices/FwkIO/plugins/DQMRootSource.cc

```diff
@@ -464,110 +464,72 @@ edm::InputSource::ItemType DQMRootSource::getNextItemType() { return m_nextItemT
-          ex.addContext("Opening DQM Root file");
-          ex << "\nInput file " << filename << " was not found, could not be opened, or is corrupted.\n";
-          throw ex;
-        }
-      }
-      originalInfo = e.additionalInfo();  // save in case of fallback error
-    }
-
-    // Check if a file is usable
-    if (file && !file->IsZombie()) {
-      logFileAction("Successfully opened file ", filename.c_str());
-    } else {
-      if (!hasFallback) {
-        if (!m_skipBadFiles) {
-          edm::Exception ex(edm::errors::FileOpenError);
-          ex << "Input file " << filename.c_str() << " could not be opened.\n";
-          ex.addContext("Opening DQM Root file");
-          throw ex;
-        }
-      }
-      if (file) {
-        delete file;
-        file = nullptr;
-      }
-    }
-
-    if (!file && hasFallback) {
-      logFileAction("  Initiating request to open fallback file ", fallbackname.c_str());
+    //loop over names of a file, each of them corresponds to a data catalog
+    bool isGoodFile(true);
+    std::vector<std::string> fNames =
+        fileitem.fileNames();  //get all names of a file, each of them corresponds to a data catalog
+    for (std::vector<std::string>::iterator it = fNames.begin(); it != fNames.end(); ++it) {
+      // Try to open a file
       try {
-        {
-          TDirectory::TContext contextEraser;
-          file = TFile::Open(fallbackname.c_str());
-        }
+        file = TFile::Open(it->c_str());
+
+        // Exception will be trapped so we pull it out ourselves
         std::exception_ptr e = edm::threadLocalException::getException();
         if (e != std::exception_ptr()) {
           edm::threadLocalException::setException(std::exception_ptr());
           std::rethrow_exception(e);
         }
+
       } catch (cms::Exception const& e) {
-        file = nullptr;  // is there anything we need to free?
-        if (m_skipBadFiles) {
-          continue;
-        } else {
-          edm::Exception ex(edm::errors::FileOpenError, "", e);
-          ex.addContext("Opening DQM Root file");
-          ex << "\nInput file " << filename << " and fallback input file " << fallbackname
-             << " were not found, could not be opened, or are corrupted.\n";
-          for (auto const& s : originalInfo) {
-            ex.addAdditionalInfo(s);
+        file = nullptr;                       // is there anything we need to free?
+        if (std::next(it) == fNames.end()) {  //last name corresponding to the last data catalog to try
+          if (!m_skipBadFiles) {
+            edm::Exception ex(edm::errors::FileOpenError, "", e);
+            ex.addContext("Opening DQM Root file");
+            ex << "\nInput file " << it->c_str() << " was not found, could not be opened, or is corrupted.\n";
+            throw ex;
           }
-          throw ex;
+          isGoodFile = false;
         }
+        originalInfo = e.additionalInfo();  // save in case of error when trying next name
```

Comment:
Yes, I agree. Here is the new implementation of DQMRootSource, which address your suggestion below as well.
https://www.dropbox.com/s/v9uvmwbby18u83w/DQMRootSource.cc?dl=0
Could you check this and let me know any change needed before I update the repos (this helps to avoid iterations of code tests/checks by jenkins)

Reply:
By quick look it looks good. (I would actually prefer to read the changes as a diff in GitHub, "code checks" is fast, tests are launched manually and so can be avoided until we're done with the visual inspection)

Reply:
@makortel, I pushed modifications to GitHub.